framework/decode/dx12_replay_consumer_base.cpp

```diff
@@ -1078,8 +1118,54 @@ HRESULT Dx12ReplayConsumerBase::OverrideD3D12CreateDevice(HRESULT
 
     IUnknown* adapter = GetCreateDeviceAdapter(adapter_info);
 
-    auto replay_result =
-        D3D12CreateDevice(adapter, minimum_feature_level, *riid.decoded_value, device->GetHandlePointer());
+    auto replay_result = E_FAIL;
+#ifdef GFXRECON_AGS_SUPPORT
+    if (options_.ags_inject_markers)
+    {
+        AGSDX12DeviceCreationParams creation_params = {};
+        creation_params.pAdapter                    = GetAdapter();
+        creation_params.iid                         = IID_ID3D12Device;
+        creation_params.FeatureLevel                = minimum_feature_level;
+
+        AGSDX12ExtensionParams extension_params = {};
+        AGSDX12ReturnedParams  returned_params  = {};
+
+        // Create AGS device for marker injection
+        AGSReturnCode ags_return = agsDriverExtensionsDX12_CreateDevice(
+            ags_marker_injector_->Context(), &creation_params, &extension_params, &returned_params);
+
+        if (ags_return == AGS_SUCCESS)
+        {
+            GFXRECON_LOG_DEBUG("Created AGS device.");
+
+            device->SetHandlePointerData(returned_params.pDevice);
```

Comment:
I'd like to avoid modifications to `HandlePointerDecoder`, but I think this could be accomplished locally with existing `HandlePointerDecoder` methods:
```
            if (device->GetHandlePointer() != nullptr)
            {
                *device->GetHandlePointer() = returned_params.pDevice;
            }
```